### DISCUSSION TITLE
question about identity model

### DISCUSSION THREAD

**Discussion by JerryZhou343** [ory/kratos#2217] [Q&A] [answered]
hey, i am fresh man for kratos. i read some code about identity module. for the identity model find these model.
- Identity 
- Credentials
- CredentialTypes
- CredentialIdentifiers
i don't understand why we need CredentialTypes model. in the struct just have two filed,
```go	// swagger:ignore
	CredentialsTypeTable struct {
		ID   uuid.UUID       `json:"-" db:"id"`
		Name CredentialsType `json:"-" db:"name"`
	}
```
and CredentialsType reference in next model
```go
	CredentialIdentifier struct {
		ID         uuid.UUID `db:"id"`
		Identifier string    `db:"identifier"`
		// IdentityCredentialsID is a helper struct field for gobuffalo.pop.
		IdentityCredentialsID uuid.UUID `json:"-" db:"identity_credential_id"`
		// IdentityCredentialsTypeID is a helper struct field for gobuffalo.pop.
		IdentityCredentialsTypeID uuid.UUID `json:"-" db:"identity_credential_type_id"`
		// CreatedAt is a helper struct field for gobuffalo.pop.
		CreatedAt time.Time `json:"created_at" db:"created_at"`
		// UpdatedAt is a helper struct field for gobuffalo.pop.
		UpdatedAt time.Time `json:"updated_at" db:"updated_at"`
		NID       uuid.UUID `json:"-"  faker:"-" db:"nid"`
	}
```
 write sql like
```go
// #nosec G201
	if err := p.GetConnection(ctx).RawQuery(fmt.Sprintf(`SELECT
    ic.identity_id
FROM %s ic
         INNER JOIN %s ict on ic.identity_credential_type_id = ict.id
         INNER JOIN %s ici on ic.id = ici.identity_credential_id
WHERE ici.identifier = ?
  AND ic.nid = ?
  AND ici.nid = ?
  AND ict.name = ?`,
		corp.ContextualizeTableName(ctx, "identity_credentials"),
		corp.ContextualizeTableName(ctx, "identity_credential_types"),
		corp.ContextualizeTableName(ctx, "identity_credential_identifiers"),
	),
		match,
		nid,
		nid,
		ct,
	).First(&find); err != nil {
		if errors.Cause(err) == sql.ErrNoRows {
			return nil, nil, sqlcon.HandleError(err) // herodot.ErrNotFound.WithTrace(err).WithReasonf(`No identity matching credentials identifier "%s" could be found.`, match)
		}

		return nil, nil, sqlcon.HandleError(err)
	}
```
i think mybe we can reduce a model element and so we can reduce a table. identity_credential_types  why we need a class to descript it?

**Comment by vinckr** [comment, marked as answer]
Hello @JerryZhou343 
I haven't worked on Kratos code, but I think this is for describing the different credential types and is set up this way to allow for the addition of more types in the future. From the [docs](https://www.ory.sh/docs/kratos/concepts/credentials):

> Ory Kratos supports several credential types:
> 
> - password: The most common identifier (username, email, ...) + password credential.
> - oidc: The "Log in with Google/Facebook/GitHub/..." credential.
> - Other credentials - support other credential types (X509 Certificates, Biometrics, ...) at will be added a later stage.
